### PR TITLE
SALTO-5983: Okta: Add deploy support for `EmailDomain`

### PR DESCRIPTION
**Pull request by Nurdok**
Add deploy support for `EmailDomain`: add, modify, remove.

---

_Additional context for reviewer_

---
_Release Notes_: 
_Okta adapter_:
- Added deploy support for `EmailDomain`
---
_User Notifications_: 
None

### CODE REVIEW

File: packages/okta-adapter/src/change_validators/email_domain_addition.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  }
+
+  return awu(changes)
+    .filter(isInstanceChange)
+    .filter(isAdditionChange)
+    .map(getChangeData)
+    .filter(instance => instance.elemID.typeName === EMAIL_DOMAIN_TYPE_NAME)
+    .filter(async emailDomainInstance =>
+      // Check if there is at least one brand that uses the email domain, return True if there isn't.
+      _.isEmpty(await findReferencingBrands(emailDomainInstance.elemID, elementsSource)),
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/email_domain_addition.ts:31]: I think we should first get all brands from the element source, and not doing this per change

File: packages/okta-adapter/src/change_validators/email_domain_addition.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    )
+    .map(instance => ({
+      elemID: instance.elemID,
+      severity: 'Error' as const,
+      message: 'Cannot add email domain without at least one brand that uses it',
+      detailedMessage: 'Cannot add email domain without at least one brand that uses it',
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/email_domain_addition.ts:37]: the `detailedMessage` should ideally give a user some instructions about how he can resolve this issue.
maybe change to something in the spirit of -
`To deploy this change, ....`

File: packages/okta-adapter/src/definitions/deploy/types/email_domain.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+export const addBrandIdToRequest: definitions.AdjustFunction<definitions.deploy.ChangeAndContext> = async ({
+  value,
+  context,
+}) => {
+  validatePlainObject(value, EMAIL_DOMAIN_TYPE_NAME)
+  const emailDomainElemId = getChangeData(context.change).elemID
+  const [brand] = (await findReferencingBrands(emailDomainElemId, context.elementSource)) ?? []
+
+  if (brand === undefined) {
+    throw new Error(`Brand not found for email domain ${emailDomainElemId.getFullName()}`)
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/definitions/deploy/types/email_domain.ts:53]: please add a `log.error` here as well

File: packages/okta-adapter/src/definitions/deploy/types/email_domain.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  }
+  // Use the ID directly instead of a reference value to avoid circular references.
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/definitions/deploy/types/email_domain.ts:55]: I think this comment is redundant

**Review by shir-reifenberg — COMMENTED** [review]
Looks great, added some comments (mainly around the usage of the `elementSource`)

File: packages/okta-adapter/src/change_validators/email_domain_addition.ts (hunk 1 above)
Comment (Nurdok) [reply packages/okta-adapter/src/change_validators/email_domain_addition.ts:31]: We do this per-change in the transform function, so if it's critical we perform this filtering only once, then maybe the transformation should happen in a filter instead so we can optimize it too?

Not sure if this is really costly in absolute time.

Comment (shir-reifenberg) [reply packages/okta-adapter/src/change_validators/email_domain_addition.ts:31]: there is actually cache in the `elementSource`, so I _think_ we should be fine here  (we can start like this and adjust if needed)
another alternative for the deploy is store the results of `findReferencingBrands` in the `sharedContext` and then call `findReferencingBrands` only if it's empty.

File: packages/okta-adapter/src/definitions/deploy/types/email_domain.ts (hunk 4 above)
Comment (Nurdok) [reply packages/okta-adapter/src/definitions/deploy/types/email_domain.ts:55]: I actually tried a reference before using the ID and found this out the hard way, so I'd rather keep it :)

**Pull request merged**